securityscanutils/securityscan.go

```diff
@@ -0,0 +1,281 @@
+		}
+	}
+	return nil
+}
+
+func (r *SecurityScanRepo) GetImagesToScan(versionToScan *semver.Version) ([]string, error) {
+	var imagesToScan []string
+	for constraintString, images := range r.Opts.ImagesPerVersion {
+		constraint, err := semver.NewConstraint(constraintString)
+		if err != nil {
+			return nil, eris.Wrapf(err, "Error with constraint %s", constraint)
+		}
+		if constraint.Check(versionToScan) {
+			// We want to make sure that each version only matches ONE constraint provided
+			// in the constraint -> []images map, so that we are scanning the right images for each version
+			if imagesToScan != nil {
+				return nil, eris.Errorf(
+					"version %s matched more than one constraint provided, please make all constraints mutually exclusive", versionToScan.String())
+			}
+			imagesToScan = images
+		}
+
+	}
+	if imagesToScan == nil {
+		return nil, eris.Errorf("version %s matched no constraints and has no images to scan", versionToScan.String())
+	}
+	return imagesToScan, nil
+}
+
+// Runs trivy scan command
+// returns if trivy scan ran successfully and error if there was one
+func RunTrivyScan(image, version, templateFile, output string) (bool, error) {
+	// Ensure Trivy is installed and on PATH
+	_, err := exec.LookPath("trivy")
+	if err != nil {
+		return false, eris.Wrap(err, "trivy is not on PATH, make sure that the trivy v0.18 is installed and on PATH")
+	}
+	args := []string{"image", "--severity", "HIGH,CRITICAL", "--format", "template", "--template", "@" + templateFile, "--output", output, image}
+	cmd := exec.Command("trivy", args...)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		// delete empty trivy output file that may have been created
+		_ = os.Remove(output)
+		// swallow error if image is not found error, so that we can continue scanning releases
+		// even if some releases failed and we didn't publish images for those releases
+		if IsImageNotFoundErr(string(out)) {
+			log.Warnf("image %s not found for version %s", image, version)
+			return false, nil
+		}
+		return false, eris.Wrapf(err, "error running trivy scan on image %s, version %s, Logs: \n%s", image, version, string(out))
+	}
+	return true, nil
+}
+
+type SarifMetadata struct {
+	Ref       string `json:"ref"`
+	CommitSha string `json:"commit_sha"`
+	Sarif     string `json:"sarif"`
+}
+
+func (r *SecurityScanRepo) UploadSecurityScanToGithub(fileName, versionTag string) error {
+	sha, err := githubutils.GetCommitForTag(r.Owner, r.Repo, "v"+versionTag, true)
+	if err != nil {
+		return err
+	}
+	githubToken, err := osutils.GetEnvE("GITHUB_TOKEN")
+	if err != nil {
+		return err
+	}
+	sarifFileBytes, err := ioutil.ReadFile(fileName)
+	if err != nil {
+		return eris.Wrapf(err, "error reading sarif file %s", fileName)
+	}
+	var sarifFile bytes.Buffer
+	gzipWriter := gzip.NewWriter(&sarifFile)
+	_, err = gzipWriter.Write(sarifFileBytes)
+	if err != nil {
+		return eris.Wrap(err, "error writing gzip file")
+	}
+	gzipWriter.Close()
+	if len(sha) != 40 {
```

Comment:
what would cause this case to occur?

Reply:
https://github.com/solo-io/go-utils/pull/455#discussion_r673579465